src/main/java/io/debezium/connector/spanner/db/metadata/SchemaRegistry.java

```diff
@@ -89,38 +94,70 @@ private boolean validate(TableId tableId, List<Column> rowType) {
         return SchemaValidator.validate(schemaTable, watchedTable, rowType);
     }
 
-    public synchronized boolean updateSchema(Timestamp updatedTimestamp) {
+    public synchronized boolean updateSchema(TableId tableId, Timestamp updatedTimestamp, List<Column> rowType) {
         if (updatedTimestamp.equals(this.timestamp)) {
             return false;
         }
 
         LOGGER.info("Schema is outdated. Try to update schema registry...");
-        forceUpdateSchema(updatedTimestamp);
+        forceUpdateSchema(tableId, updatedTimestamp, rowType);
         LOGGER.info("Schema registry has been updated to date {}", updatedTimestamp);
         return true;
     }
 
-    private void forceUpdateSchema(Timestamp updatedTimestamp) {
-        this.timestamp = updatedTimestamp;
-        this.changeStream = schemaDao.getStream(timestamp, streamName);
+    // Used for unit testing only to set the mock SpannerSchema class
+    @VisibleForTesting
+    void setSpannerSchema(SpannerSchema spannerSchema) {
+        this.spannerSchema = spannerSchema;
+    }
 
-        if (this.changeStream == null) {
-            throw new IllegalStateException("Change stream doesn't exist at timestamp: "
-                    + streamName + ", " + timestamp);
-        }
+    public void updateSchemaFromStaleTimestamp(TableId tableId, Timestamp timestamp, List<Column> rowType) {
+        LOGGER.info("Schema is outdated. Try to update schema registry outside of retention period...");
+        TableSchema watchedTable = this.getWatchedTable(tableId);
```

Comment:
Can you get the table name from the TableId, instead of making a call to "getWatchedTable"? https://github.com/debezium/debezium-connector-spanner/blob/main/src/main/java/io/debezium/connector/spanner/db/metadata/TableId.java#L30

At the same time, you should remove the mock to getWatchedTable from the unittest